In [2]:
from IPython.core.display import HTML
def css_styling():
    styles = open("./styles/custom.css", "r").read()
    return HTML(styles)
css_styling()

### BEFORE YOU DO ANYTHING...
In the terminal:
1. Navigate to __inside__ your ILAS_Python repository.
2. __COMMIT__ any un-commited work on your personal computer.
3. __PULL__ any changes *you* have made using another computer.
4. __PULL__ textbook updates (including homework answers).

Then:
1. __Open Jupyter notebook:__   Start >> Programs (すべてのプログラム) >> Programming >> Anaconda3 >> JupyterNotebook
1. __Navigate to the ILAS_Python folder__. 
1. __Open today's seminar__  by clicking on 11_Testing.

Some packages we will use today...

In [37]:
import numpy as np
import sys

# Testing

# Lesson Goal
To use a standard format to tests the code you have written against possible inputs. 

# Objectives

- Learn the fundamentals of writing and running tests
- Understand the effect of floating point number storage on comparison operations and how to deal with this in your code.
- Learn how to structure your file system and `import`s to keep tests in seperate files from the main program. 

# Why we are studying this

The programming examples in the preceding notebooks included little or no checking/testing. 

TDD (test driven development) is a successful and widely-accepted method of developing code.

Testing is a critical part of software engineering:
 - enhancing program quality.
 - builds the confidence we (and others) have in a program to run.

# Lesson Structure
 - Introduction
 - The `assert` command.
 - Encapsulating Tests.
 - Test frameworks: `pytest`
 - Designing tests: two worked examples.
 - Testing in scientific computing:
  - Floating point number storage
  - Comparing floating point numbers using `Numpy`
  - Writing tests for floating point numbers
 - Importing your program to a *suite* of tests
 

# Introduction to Testing

Testing is useful in developing a new program. 

Tests are code that is written to check the program performs exactly as you expect it to. 

Programs with dynamically changing input should also come with a *suite* of tests that can be run regularly.

For example, any program used for engineering analysis should have a suite of tests.

The data input by the user or imported, for example from the results of an experiment, may vary so the program's ability to repspond to this should be tested.

When testing a program, we should test for:
 - valid input data. 
 - invalid input data.
 
For the valid cases the computed result should be checked against a known correct solution. 

For the invalid data cases, tests should check that an exception is raised. 

## `assert` : Checking your code against a *known* solution.

Let's look again at an example from Seminar 5: Functions; a recursive function used to generate the Fibonacci number series; an integer sequence characterised by the fact that every number (after the first two) is the sum of the two preceding numbers in the sequence. 

$$
f_n = f_{n-1} + f_{n-2}
$$

In [1]:
def f(n): 
    "Compute the nth Fibonacci number using recursion"
    if n == 0:
        return 0  # Breaks out of the recursion loop
    elif n == 1:
        return 1  # Breaks out of the recursion loop
    else:
        return f(n - 1) + f(n - 2)  # Calls f for n-1 and n-2 (recursion)    

We can check a number of computed terms in the series against
known results. 

| Term | 0  | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Output | 0 | 1 | 1 | 2 | 3 | 5 | 8 | 13 | 21 | 34 | 55 | 89 | 144 | 233 | 377 | 610 |

A useful tool for this is the `assert` keyword:

In [4]:
assert f(0) == 0
assert f(1) == 1
assert f(2) == 1
assert f(3) == 2
assert f(10) == 55
assert f(15) == 610

If all the assertions are true, there should be no output. 

__Try it yourself__
<br>Try changing one of the checks to raise an `AssertionError`.

## Error messages
An error message can be added to the `AssertionError`.

The error message is printed if the `assertion` fails.

For example, in the cell below, try changing 610 to another number to rasie the exception.

In [18]:
assert f(15) == 610, "Fail: Fibonacci number, term 15, is incorrect"

## Encapsulating Tests
Placing the tests within the main body of a program can be visually crowding.

It can also be time-consuming and unecessary to run the tests *every* time the code is executed.

Tests, like other sections of code, can be placed within __functions__ to improve the tidiness and readablity of the program.



For example, several `assert` statements may be placed within a function, allowing them to be called using one line.



In [19]:
def test_fibonacci():
    assert f(0) == 0
    assert f(1) == 1
    assert f(2) == 1
    assert f(3) == 2
    assert f(10) == 55
    assert f(15) == 610, "Fail: Fibonacci number, term 15, is incorrect"
    
test_fibonacci()

## Test frameworks 

Testing is so important to good software engineering that many packages have been developed to help write and run tests. 



One of the most useful features is the ability to switch between running your code:
- with tests (while developing and periodically testing a program)
- without tests (for normal operation)

### `pytest`
Today we will learn to use a popular and powerful testing package for Python called `pytest` (http://doc.pytest.org/en/latest/). 

Before using `pytest`, we need to check that it is installed. 

Do do this we can use `try` and `except` (Seminar 10: Error Handling)

We try to import `pytest`, and if that fails we install it:

In [20]:
try: 
    import pytest
except:
    try:
        !{sys.executable} -m pip -q install pytest
        import pytest
    except ImportError:
        !{sys.executable} -m pip -q --user install pytest

Executing a program using `pytest`:
- runs the code
- calls any testing functions (the only requirement is that the function name starts with `test_` to be detected as a test)

By using `pytest`, testing functions do not need to be *called* (and potentially later removed from) within the main body of the code.





Let's look at an example.

You can find the file __`test_fib_example.py`__ within the ILAS_python repository
<br>(`ILAS_python/sample_data/pytest_examples/test_fib_example.py`).

The file contains the following code...



```Python
    import pytest

    def f(n): 
        "Compute the nth Fibonacci number using recursion"
        if n == 0:
            return 0  # Breaks out of the recursion loop
        elif n == 1:
            return 1  # Breaks out of the recursion loop
        else:
            return f(n - 1) + f(n - 2)  # Calls f for n-1 and n-2 (recursion) 

    def test_fibonacci():
        assert f(0) == 0
        assert f(1) == 1
        assert f(2) == 1
        assert f(3) == 2
        assert f(10) == 55
        assert f(15) == 610, "Fail: Fibonacci number, term 15, is incorrect"
        ```

The code may be executed from the command line, <br>(from within ILAS_python/sample_data/pytest_examples/) using:
> `python3 test_fib_example.py`

To run `pytest` instead:
>`pytest test_fib_example.py`

A message is shown to show that a test has been passed.

    =========================== 1 passed in 0.01 seconds ===========================

To run `pytest` in *verbose* mode use `-v`:
> `pytest -v test_fib_example.py`

This returns more information, for example the name of the the function tht passed.

    test_fib_example.py::test_fibonacci PASSED

    =========================== 1 passed in 0.02 seconds ===========================

To run all tests in all files in a directory:
>`pytest *` 

## Designing a test: Calculator Example
The general idea of *test driven development* (using tests to build and improve programs) is to:
- write a test that fails
- make the test pass 
- refactor (edit your code, keeping the same functionality)

Let's look at a very simple example.

You can find the file __`calculator_example.py`__ within the ILAS_python repository
<br>(`ILAS_python/sample_data/pytest_examples/calculator_example.py`).

The file contains the following code (and some commented code):
```Python
    def test_calculator_add_returns_correct_result():
            result = calc_add(2,2)
            assert result == 4
            ```

At first the test fails because there is no function with the name `calc_add` defined in the file.

    =================================== FAILURES ===================================
    __________________ test_calculator_add_returns_correct_result __________________

    def test_calculator_add_returns_correct_result():
    >   	result = calc_add(2,2)
    E    NameError: name 'calc_add' is not defined



Let's start by adding a function that does nothing.
```Python
    def calc_add(x,y):
        pass
        ```

This time the code fails as we expect.

`pytest` shows the line causing the test to fail.  

    =================================== FAILURES ===================================
    __________________ test_calculator_add_returns_correct_result __________________

    def test_calculator_add_returns_correct_result():
    	result = calc_add(2,2)
    >   	assert result == 4
    E    assert None == 4




Let's fix the function and see if our test passes now:
```Python
    def calc_add(x,y):
        # pass
        
        return x + y
        ```

We have defined `calc_add` and it works as expected. 

However, we have only tested the case we are interested in at the moment.

There is more work needed to ensure the function has been fully tested.



What would happen if a user called the function, giving non-numeric inputs as arguments?

Python allows for the addition of strings and other types.

However, the calculator will only work correctly (add two numbers) if numeric inputs are given. 



We can __test__ that an exception is raised when varibales of non-numeric type are entered using `try`, `except` and `assert`:
```Python
    def test_calculator_returns_error_message_if_both_args_not_numbers():

        try:
            calc_add("two", "three")


        except ValueError:
            print("Exception caught")
            assert True, "Fail: ValueError exception not caught"


        except:
            assert False, "Fail: Exception other than ValueError caught"


        else: 
            assert False, "Fail: No exception caught"
            ```

If we are only interested in detecting the `ValueError` we can write:
are rasied.
```Python
    try:
        calc_add("two", "three")

    except ValueError:
        print("Exception caught")
        assert True, "Fail: ValueError exception not caught"
```
far more concisely using a `pytest` method.
<br>`pytest.raises()` is dedicated to testing wether exceptions are rasied.
```Python        

    with pytest.raises(ValueError):
        calc_add("two", "three")
```

The test fails, indicating that we are not raising the ValueError when we expect to be. 

We now have a new failing test. 

Next, we code the solution to make it pass.

To check only numeric values are entered, we need to write an exception.

In this example, the built-in function `isinstance` is used to check the data type of a variable against a set of permissible types: 


```Python
number_types = (int, float, complex)

    def calc_add(x,y):
        # pass
        
        # return x+y
 
        if isinstance(x, number_types) and isinstance(y, number_types):
            return x + y
        else:
            raise ValueError("Non-numeric input given")
            ```

The test now passes.



To complete the testing, the aim is to add test cases to cover all scenarios.

For example, as there are two variables, it means that either could potentially not be a numeric value. 

A full set of tests is shown below:


In [21]:
def test_calculator_add_method_returns_correct_result():
    result = calc_add(2,2)
    assert result == 4

def test_calculator_returns_error_message_if_both_args_not_numbers():
     with pytest.raises(ValueError):
            calc_add("two", "three")
            
def test_calculator_returns_error_message_if_x_arg_not_number():
     with pytest.raises(ValueError):
            calc_add("two", 3)

def test_calculator_returns_error_message_if_y_arg_not_number( ):
     with pytest.raises(ValueError):
            calc_add(2, "three")

When we run all the tests, we can confirm that the function `calc_add` meets our requirements.

## Designing a test: Parameter Validity Checking Example

Now let's look at the hydrostatic pressure function that we studied in:
 - Seminar 5: Functions
 - Seminar 10: Error Handling; where we studied how to raise an exception if the arguments to the function were invalid.

In [2]:
def hp(h, *, rho = 1000, g = 9.81):
    """
    Computes the hydrostatic pressure acting on a submerged object given:
        - the height of fluid above the object, h
        - the density of the fluid in which is it submerged, rho
        - the acceleration due to gravity, g

    """
    if h < 0:
        raise ValueError("Height of fluid, h, must be greater than or equal to zero")
    if rho < 0:
        raise ValueError("Density of fluid, rho, must be greater than or equal to zero")
    if g < 0:
        raise ValueError("Acceleration due to gravity, g, must be greater than or equal to zero")

    return rho * g * h


When we add a new feature to a program, we should test it.

In this case we should test that  a `ValueError` error *is*  raised if invalid data is input.

You can find the file __`hydrostatic_pressure_example.py`__ within the ILAS_python repository
<br>(`ILAS_python/sample_data/pytest_examples/calculator_example.py`).



The file contains the function `hp` and three tests to check the validity of each function argument; `h`, `rho` and `g`:
```Python

import pytest

def test_hydrostatic_pressure_returns_error_if_h_less_then_0():
    with pytest.raises(ValueError):
        hp(-10)

def test_hydrostatic_pressure_returns_error_if_rho_less_then_0():
    with pytest.raises(ValueError):
        hp(10, rho=-10)

def test_hydrostatic_pressure_returns_error_if_g_less_then_0():
    with pytest.raises(ValueError):
        hp(10, g=-9.81)
        ```

## Testing in Scientific Computing
There are two issues when testing scientific computing software. 

Direct comparison of:
1. floats 
1. arrays

is not possible

###  Floating point storage

Most engineering calculations involve floating point numbers. 

Computers store floating point numbers by storing the sign, the significand (also known as the mantissa) and the exponent, e.g.: for $10.45$

$$
10.45 = \underbrace{+}_{\text{sign}} \underbrace{1045}_{\text{significand}} \times \underbrace{10^{-2}}_{\text{exponent} = -2}
$$



Python uses 64 bits to store a `float` (in C and C++ this is known as a `double`). 

One bit is used to store the sign.

52 bits are used for the significance.

11 bits are used for the exponent. 

The precision with which numbers can be represented is limited by the number of bits used to store the number.

A 64 bits a floating point number is usually __precise__ to 15 to 17 significant figures.

The internal representation of floats, introduces minute errors to numerical values. 

$$
10.45 = \underbrace{+}_{\text{sign}} \underbrace{1045}_{\text{significand}} \times \underbrace{10^{-2}}_{\text{exponent}}
$$

This is because computers use base 2 binary representation (not base 10, for example) to store the significance and the exponent.

Some fractions cannot be represented exactly by computers in base 2.

$1/10 = 0.1$ 
<br>cannot be represented exactly in binary. 

$1/2 = 0.5 = 2^{-1}$ <br>can be represented exactly in binary. 

To demonstrate, let's assign the number 0.1 to the variable `x`:

In [22]:
print(0.1)

0.1


This looks fine, but the `print` statement is hiding some details. 

We can use a user-defined number of decimal places to represent a number using the `str.format` method.



In [23]:
print('{0:.30f}'.format(0.1))

0.100000000000000005551115123126


"Modulo-formatting" can also be used and is suitable for older versions of Python (Python 2).

In [24]:
print('%.30f' % 0.1)

0.100000000000000005551115123126


Using 30 characters shows an inexact representation of 0.1.

The stored value is the nearest representable binary fraction to 0.1.

The difference between 0.1 and the binary representation is known as the *roundoff error*.

We can see that the representation is accurate to about 17 significant figures.



Interestingly, there are sometimes different decimal numbers that share the same nearest approximate binary fraction. 

The numbers 
- 0.1
- 0.10000000000000001
- 0.1000000000000000055511151231257827021181583404541015625 

are all approximated by:

3602879701896397 / 2 ** 55. 

Since all of these decimal values share the same approximation, any one of them could be displayed.  

Checking for 0.5, we see that it appears to be represented exactly:`

In [26]:
print('{0:.30f}'.format(0.5))

0.500000000000000000000000000000


The round-off error for the 0.1 case is small.

In many cases will not present a problem. 

However, sometimes accumulation of round-off errors can lead to significant inaccuracies. 

### Example: Inexact Representation of Floating Point Numbers

Consider the expression

$$
x = 11x - 10x
$$



$10x = 1$, where $x = 0.1$

So if $x = 0.1$ we can  write the expression above as:

$$
x = 11x - 1
$$



Now, starting with $x = 0.1$ we evaluate the right-hand side to get a 'new' $x$.

The 'new' $x$ should be equal to the original value of $x$.

In other words, $x$ should remain equal to $0.1$.



However, if we repeat this process 20 times the solution blows up, deviating widely from $x = 0.1$. 

In [27]:
x = 0.1

for i in range(20):
    x = x*11 - 1
    print(x)

0.10000000000000009
0.10000000000000098
0.10000000000001075
0.10000000000011822
0.10000000000130038
0.1000000000143042
0.10000000015734622
0.10000000173080847
0.10000001903889322
0.10000020942782539
0.10000230370607932
0.10002534076687253
0.10027874843559781
0.1030662327915759
0.13372856070733485
0.4710141677806834
4.181155845587517
44.992714301462684
493.9198573160895
5432.118430476985


If we repeat this operation for $x = 0.5$, we see that the result is exact.

In [28]:
x = 0.5

for i in range(20):
    x = x*11 - 5
    print(x)

0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5


By default, Python uses 64 bits to store a float. 

We can use the module NumPy to cast a float as a 32 bit number. 

Testing this for the $x = 0.1$ case:

In [29]:
x = np.float32(0.1)

for i in range(20):
    x = x*11 - 1
    print(x)

0.100000016391
0.100000180304
0.100001983345
0.10002181679
0.100239984691
0.102639831603
0.129038147628
0.419419623911
3.61361586303
38.7497744933
425.247519426
4676.72271369
51442.9498506
565871.448356
6224584.93192
68470433.2511
753174764.762
8284922411.38
91134146524.2
1.00247561177e+12


The error blows up faster in this case compared to the 64 bit case.

Using 32 bits leads to a poorer approximation of $0.1$ than when using 64 bits.

## Comparing floating point numbers using Numpy
Due to this inherent inaccuracy, the equality of floating point numbers cannot be compared using regular comparison operators.



In [30]:
0.1 + 0.2 == 0.3

False

For example, the following test would fail:

```Python
def test_calculator_add_method_returns_correct_result(self):
    result = calc_add(0.1,0.2)
    assert result == 0.3
    ```

The `Numpy` function `allclose` provides a way to evaluate this comparison:

In [31]:
np.allclose(0.1 + 0.2, 0.3)

True

## Comparing arrays using Numpy
An array can contain multiple values. 

The boolean `True` or `False` outcome of the conditional can potentially vary from element to element. 

In [33]:
A = np.array([1.,2.])

if A == A: 
    pass

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

The `numpy.allclose` function is also used to specify that all values of an array are being considered. 

In [34]:
np.allclose(A, A)

True

## Writing tests for floating point numbers.

`allclose` should be used to test equality of floats or matrices. 

For example, let's add the following function to calculator_example.py

```Python
def test_calculator_add_method_returns_correct_result_for_floats(self):
    result = calc_add(0.1,0.2)
    assert np.allclose(result, 0.3)
    ```

### The `Numpy` testing module.

A more concise way to write this is using `Numpy`'s dedicated testing module.

```Python

import numpy.testing as npt

```



The tests for the comparisons above can be written:

```Python

npt.assert_allclose(0.1 + 0.2, 0.3)
npt.assert_allclose(A, A)
```

## Importing your program to a *suite* of tests.

As you write more and more tests, it can be beneficial to organise your code so that your tests are stored in a seperate file or directory.

There are rules for running files using `pytest` that depend on the position of the file containing the tests, relative to the file containing the program, in the computer's file system.

`pytest` supports three common test layouts:

### 1) Same directory for tests and program files 

    directory/
        program.py
        tests.py
        
######  `tests.py` must contain :

```Python
import pytest
import program
from program import *
```

 



###### To run the tests:
- In the terminal, use `cd` to navigate to within the directory containing both files.
- run:<br>`pytest tests.py`

### 2) Inline test directories

    main_directory/
    
        directory/

            __init__.py

            program.py

            tests/
                tests.py
                




###### `tests.py` must contain :

```Python
import pytest
import sys
import os

myPath = os.path.dirname(os.path.abspath(__file__))

# The full path to the file is specified.
# The 0 makes the next directory up the first place python looks. (important if a file of the name name exists elsewhere in directory).

sys.path.insert(0, myPath + '/../')

import program
from program import *
```

###### To run the tests:

In the terminal, from within `main_directory`:

     main_directory/
            directory/
                __init__.py
                program.py
                tests/
                    tests.py
                
                
- `pytest directory/tests/tests.py`   <br>to run tests in one file

- `pytest *` <br>to run all tests below current directory
 
- `pytest directory/*` <br>to run all tests below directory

- `pytest directory/tests/* ` <br>to run all tests below directory/tests



In the terminal, from within `directory`:

     main_directory/
            directory/
                __init__.py
                program.py
                tests/
                    tests.py
                    
- `pytest tests/tests.py`   <br>to run tests in one file

- `pytest *` <br>to run all tests below current directory

- `pytest tests/* ` <br>to run all tests below directory/tests


In other words you specify the path to the tests from wherever you are running `pytest`.

This allows you to choose which tests you run and when. 

### 3) Seperate directories for program and tests


    directory
    program/
        __init__.py
        program.py

    tests/
        tests.py

###### `tests.py` must contain :

```Python
import pytest
import sys
import os

myPath = os.path.dirname(os.path.abspath(__file__))

# The full path to the file is specified.
# The 0 makes the next directory up the first place python looks. (important if a file of the name name exists elsewhere in directory).

sys.path.insert(0, myPath + '/../tests')

import program
from program import *
```

###### To run the tests:

In the terminal, from within `directory`:

    directory
        program/
            __init__.py
            program.py
        tests/
            tests.py
                
- `pytest tests/tests.py`   <br>to run tests in one file

- `pytest *` <br>to run all tests below current directory

- `pytest tests/* ` <br>to run all tests below directory/tests



# Summary

 - Tests are code that is written to check the program performs exactly as you expect it to. 
 - The `assert` command can be used to check your code againsta  known solution.
 - Tests are usually placed within __functions__ to improve tidiness, readablity and to allow them to be recognised by packages such as `pytest`.
 The general format of *test driven development* is:
  - write a test that fails
  - make the test pass 
  - refactor
 - `with pytest.raises` can be used to test for exceptions.
 - Due to the way that computers store floating point numbers, it is not possible to directly compare:
    1. floats 
    1. arrays
 - The method `Numpy.allclose` can be used to compare floating point numbers. 
 - Numpy has its own testing module; `Numpy.testing` which makes writing tests on floting point numbers more concise.
 - Code can be made neater and more readable by storing your tests in a seperate file to your main program. 